### ISSUE TITLE
Support C++ modules

### ISSUE THREAD

**Issue by abergmeier** [bazelbuild/bazel#4005]
### Description of the problem / feature request / question:

Starting to look into C++ Modules. While it will earliest officially arrive in 2020, would be good to have support for that soonish.
Heard there is some Module support inside of Google. Can/Will that be opensourced?

**Comment by mhlopko** [comment]
Hi Andreas,
everything regarding modules (only for clang) is open sourced already. The only missing piece is a CROSSTOOL that makes use of them and provides all necessary features. But I think it would be quite a challenge to write such CROSSTOOL without understanding the implementation. If I don't find a good movie on my flight to bazel conference I might take a look and hack something together. No promises though, since this would be a free time exercise, not something that is a priority for bazel right now.

**Comment by abergmeier** [comment]
Interesting.
@mhlopko Can you perhaps elaborate a bit, what parts of CROSSTOOL would interact with modules? Personally I thought it would be more in the realm of cc-providers and additional actions for module validity checks.

**Comment by mhlopko** [comment]
Yeah you are right, the implementation does deal with providers and actions, but all that logic is already in bazel and ready to be used. What is missing is a bunch of features (and maybe action_configs, I'm not sure) in the CROSSTOOL to pass the right flags for the right actions. Names of the features are hardcoded in bazel, so they need to match.

**Comment by mhlopko** [comment]
So in other words, it's all there, just not very well documented.

**Comment by abergmeier** [comment]
> just not very well documented

Like all else about CC toolchains you mean ;)

Thanks. I am trying to get a custom hermetic Clang toolchain running anyways. Perhaps will then try this out a bit.

**Comment by mhlopko** [comment]
Exactly :) I hope I'll see you next week on the crosstool breakout session during bazel conference? :) Btw, just wrote this yesterday, I really hope I'm not one day late: https://github.com/bazelbuild/bazel/wiki/Yet-Another-CROSSTOOL-Writing-Tutorial

Anyway, don't hesitate to ask if you get stuck on the way. I'm also grateful for any input/comments/suggestions.

**Comment by abergmeier** [comment]
Since I do have problems with the include paths for libc (https://stackoverflow.com/questions/47078520/clang-toolchain-fails-for-system-libc-files), how does `libc_top` (`inputsForLibc`) work?

Especially since I cannot seem to be able to glob system files.

**Comment by ArekSredzki** [comment]
Now that C++20 Modules are nearing standardization, could we get some updates on the status of properly supporting them within Bazel?

I was fortunate to attend CppCon this year and was excited to see that many build systems were making progress on supporting modules. Unfortunately, myself and other Bazel users that I spoke to were concerned that there was no Bazel-specific information on the topic.

**Comment by hlopko** [comment]
We don't plan to work on C++ modules in 2019. We are discussing working on them in 2020 mostly in the context of objc and swift rules. We'll update this issue once we have more precise plans.

**Comment by ArekSredzki** [comment]
Thanks @hlopko, I'm sure that much of the C++ Bazel community will very much value support for them. Looking forward to hearing more.

**Comment by ctrysbita** [comment]
Any update here? :D

**Comment by ArekSredzki** [comment]
@hlopko Is there anyone that we could ping regarding the timeline for supporting C++20 modules?
There appears to be significant community interest in this feature, it'd be good to set expectations :)

**Comment by hlopko** [comment]
@oquenchil and @c-mita are authorities there.

**Comment by ArekSredzki** [comment]
@oquenchil @c-mita Bump on the above? :)

**Comment by oquenchil** [comment]
It might be that this gathers steam at the beginning of next year but I can't promise anything. In any case, it would be clang modules, not C++20 modules.

Whoever is interested in clang modules, definitely thumbs up the original issue above and that might help with prioritization.

**Comment by technoir42** [comment]
> it would be clang modules, not C++20 modules.
Should we create a separate issue for C++20 modules then? I'm only interested in those as I'm using MSVC on Windows.

**Comment by oquenchil** [comment]
Yes, you can create a separate issue for tracking C++20 module support. Be aware though that as it looks right now it will be a long time before we work on it, probably not next year.

**Comment by thii** [comment]
Can you rename this issue to "Support Clang modules"?

**Comment by rnburn** [comment]
>  In any case, it would be clang modules, not C++20 modules.

@oquenchil - By "Clang modules", do you mean Clang's non-standard [module system](https://clang.llvm.org/docs/Modules.html) that pre-dates C++20 modules? 

Why target that instead of the  C++20 modules? Clang supports a lot of the standard module spec already ([blog post](https://mariusbancila.ro/blog/2020/05/15/modules-in-clang-11/)) -- I expect enough that people could start using it.

**Comment by oquenchil** [comment]
A lot of the implementation for Clang modules already exists for the Google internal version of Bazel, the effort required to implement Clang modules in Bazel would be less than for C++20 modules. We are not discarding ever implementing the latter though, just that it would come at a later time.

**Comment by rnburn** [comment]
But why hijack this issue into a clang module feature request? The original poster clearly worded it for C++20 modules and I expect that's what most people following it want.

If someone's looking for clang's homegrown module system, can we let them submit a different issue for it? Is it even broadly used outside of Google?

C++20 module support may not be on Google's short-term roadmap, but there are a lot of project that would benefit from being able to use standard modules, so perhaps someone else might sponsor the work.

**Comment by ArekSredzki** [comment]
I second that C++20 module support would be far more beneficial to the community at this point.
Many organizations use multiple toolchains, in which case, supporting Clang's module implementation brings little value.

I suspect that most people following this ticket are interested in C++20 modules since that's what it was created for.
Could a separate ticket be created for Clang modules instead of renaming this one as that's a separate request with a niche application? 🙏

**Comment by ArekSredzki** [comment]
Thanks!

**Comment by det** [comment]
Hello! I took a shot at implementing Clang Modules in a local bazel repo as this ticket said that everything for modules is already open-sourced. I believe I have implemented most of the crosstool features nearly correct (but maybe not?) and I am running into a couple issues:
1) I'm not sure how module maps for the C++ stdlib and sysroot should be handled:
   - You must provide a `crosstool.cppmap` as the `module_map` for your `cc_toolchain`. It's not clear if this should include maps for both your c++ stdlib and sysroot.
   - Bazel doesn't ever seem to compile this module map into a pcm file. How are we supposed to acheive this?
       - Are you supposed to precompile a .pcm file and include it in your compiler files and add `-fmodule-file=path/to/crosstool.pcm` to the `use_header_modules` feature?
           - Can this even work since pcm files are sensitive to compile flags which can be changed on the command line?
2) Bazel seems to support 2 modes of modules support, with and without modules codegen. When using `--features=header_module_codegen` it will correctly try to compile the .pcm file to a .o file but it is not making available all the .pcm dependencies of the .pcm file that is being compiled. For example, if `foo.pcm` imports `bar.pcm` but when `foo.pcm` is being compiled to `foo.pcm.o` it will try to open `bar.pcm` which is not available in the sandbox and so will error. I also verified with aquery that the dependent pcm files are not provided as inputs to that compile action.

Here is what I have come up with so far for how to fill out the features: https://gist.github.com/det/46724828d1143332b29c0642691b24a3

Except for the layering check, there is no documentation or examples to be found anywhere on the internet so I would appreciate any pointers about anything I could be doing wrong as well as clarification of if clang modules should currently be usable in the open source bazel. @hlopko @oquenchil

I am using bazel 3.7.1 and Clang 11.0.0 in my toolchain if that matters.

Thanks!

**Comment by EvanHonnold** [comment]
@hlopko @oquenchil  Might you have a chance to take a look at @det 's comment above? I am running into many of the same issues and would be curious to know if there are any workarounds.

**Comment by hlopko** [comment]
I haven't yet, @c-mita, maybe you have worked with Bazel modules recently?

**Comment by oquenchil** [comment]
> Hello! I took a shot at implementing Clang Modules in a local bazel repo as this ticket said that everything for modules is already open-sourced. I believe I have implemented most of the crosstool features nearly correct (but maybe not?) and I am running into a couple issues:
> 
> 1. I'm not sure how module maps for the C++ stdlib and sysroot (libc) should be handled:
>    
>    * You must provide a `crosstool.cppmap` as the `module_map` for your `cc_toolchain`. It's not clear if this should include maps for both your c++ stdlib and sysroot.
>    * Bazel doesn't ever seem to compile this module map into a pcm file. How are we supposed to acheive this?
>      
>      * Are you supposed to precompile a .pcm file and include it in your compiler files and add `-fmodule-file=path/to/crosstool.pcm` to the `use_header_modules` feature?
>        
>        * Can this even work since pcm files are sensitive to compile flags which can be changed on the command line?
> 2. Bazel seems to support 2 modes of modules support, with and without modules codegen. When using `--features=header_module_codegen` it will correctly try to compile the .pcm file to a .o file but it is not making available all the .pcm dependencies of the .pcm file that is being compiled. For example, if `foo.pcm` imports `bar.pcm` but when `foo.pcm` is being compiled to `foo.pcm.o` it will try to open `bar.pcm` which is not available in the sandbox and so will error. I also verified with aquery that the dependent pcm files are not provided as inputs to that compile action.
> 
> Except for the layering check, there is no documentation or examples to be found anywhere on the internet so I would appreciate any pointers as well as clarification of whether clang modules should currently be usable in the open source bazel. @hlopko @oquenchil
> 
> I am using bazel 3.7.1 and Clang 11.0.0 in my toolchain if that matters.
> 
> Thanks!

So I tried this out and I get a toy example to work with modules. I added the features to unix_cc_toolchain_config.bzl and made sure I was using clang with CC=clang.  The module maps for system libraries should already be taken care of thanks to [this](https://cs.opensource.google/bazel/bazel/+/master:tools/cpp/unix_cc_configure.bzl;drc=47d231707fdd0fd1a913e34cad9c4f1dc63d33ea;l=443).

You said that you are using 3.7.1, so I suspect the problem is that you didn't re-build bazel after making changes to unix_cc_toolchain_config.bzl. You have to re-build and use the custom binary. Please try that and let me know if you bump into any issues.

**Comment by rnburn** [comment]
I put together a project with bazel rules for working with C++20 modules: 

https://github.com/rnburn/rules_cc_module

Here are some examples of how to use it

https://github.com/rnburn/rules_cc_module/tree/main/example

It requires a recent version of gcc and there are still some pieces to fill in; but I think with a bit more work, it could be made to handle most use cases. Let me know if anyone is interested in helping to finish the project.

**Comment by oquenchil** [comment]
That's interesting. Thanks for doing that Ryan.

Would you be interested in investigating how to add support for C++20 modules to the existing native rules? It would probably require a design doc and discussion.

**Comment by rnburn** [comment]
I actually think a first iteration for module support would be pretty easy.

1. Add a new rule `cc_module` that's defined something like this
```
cc_module(
      name = "module_a",
      module_name,  # specifies the module naame, defaults to <name>
      src = "module_a.cc", # required, specifies the source file with "export module <module_name>"
      impl_srcs = [
           "a_impl1.cc",    
           ...   # optional list of implementation sources, they would specify "module <module_name>" with no export keyword
      ],
     deps = [
           # list of either modules or cc_libraries
     ],
)
```
cc_module would produce both a "compiled module interface" (e.g. module_a.cmi), something built from only `src` and a standard archive (e.g. libmodule_a.a), formed from the object files for `src` and the `impl_src`'s.

2. The other cc_* attributes would need to be made module-aware: For any cc_modules that they depend on, the compiler would need to be passed a mapping telling it where the CMIs are
```
<module1> <module_cmi1>
<module2> <module_cmi2>
... 
```
For gcc this can be done with the `-fmodule-mapper=<map_file>` argument; and for clang, this can be done with `-fmodule-file=<module_name>=<module_cmi>` arguments.

There are some other features you'd want to add at some point, like

* Support for producing header unit modules
* Support for module partitions (could be done by adding partition argument to cc_module)
* Maybe support for bundling multiple modules in a single rule
* Figure out what to do if you try to mix cc_modules compiled with options that would make them incompatible (e.g. `-fno-exceptions`)

But I don't think you need to worry about those for a first cut.

**Comment by oquenchil** [comment]
Is a new rule absolutely necessary?

Support for clang modules is in the Bazel [codebase](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java), just not wired up, only for the internal version. For clang modules we don't need a new rule and we automatically generate maps. I was hoping that something similar could be done for C++20 modules and that we didn't have to write a new rule.

If C++20 modules used most of the existing codepaths in the Bazel implementation for clang modules but just differed in the crosstool then that would be the best outcome.

**Comment by rnburn** [comment]
> Support for clang modules is in the Bazel codebase, just not wired up, only for the internal version. For clang modules we don't need a new rule and we automatically generate maps. I was hoping that something similar could be done for C++20 modules and that we didn't have to write a new rule.

The concept of clang's module maps, where you express how header files are mapped to modules, really doesn't exist with C++20 modules. C++20 does have header unit modules, but they are explicit. For example, if I might write a module like this
```c++
module;
// global module fragment (can include header files)
// For background, see
// https://vector-of-bool.github.io/2019/10/07/modules-3.html
#include <cmath>
export module A;
// #include <vector> <-- this would be illegal
import "path/to/hdr.h"  
  // allowed to use a header unit module, but you need to explicitly express that you
  // want to use a header unit module with import.
  // Also, you need to set up the generation of the header unit module
```
What clang does where it implicitly translates #includes to use modules isn't part of c++20 modules and I don't know that you would gain much by trying to reuse the existing clang module machinery.

**Comment by rnburn** [comment]
> Is a new rule absolutely necessary?

You could try to reuse cc_library, but I think it's cleaner to add a new rule. For each module, there is one source fille with `export module <name>` that generates the compiled module artifact. If you were to add attributes to cc_library to express which modules get generated and which source files generate those modules, I think things would get messy.

**Comment by bjacklyn** [comment]
> So I tried this out and I get a toy example to work with modules. I added the features to unix_cc_toolchain_config.bzl and made sure I was using clang with CC=clang.

@oquenchil would you be able to post a gist for the missing cc_toolchain features that are needed to try out clang modules (not c++ 20 modules)? It looks like there are a few -- `use_header_modules`, `header_module_compile`, `header_module_codegen` -- that I wasn't able to find in the toolchain config but exist in the java code. And with those features added, is it as simple as `CC=clang` and adding `build --features=use_header_modules` to `.bazelrc` to try it out?

**Comment by dpeter99** [comment]
So from reading this issue am I correct in assuming that currently there is no way to use cpp20 Modules in bazel?
If so is there any chance we will get support in the near future?

**Comment by dpeter99** [comment]
> You'll see support exactly when googling switches to C++20 internally

I know this. what I'm asking is if there is some way to use modules, like community-driven rules. Or if not that at least some update on when if ever we can expect Google to switch.
I'm asking because I'm currently setting up a new project that I want to use a monorepo style set up for, but I also would like to use cpp20 modules.

**Comment by MahmoodMahmood** [comment]
Any ETA on support for this?

**Comment by Genomorf** [comment]
Any update here?

**Comment by oquenchil** [comment]
>@oquenchil would you be able to post a gist for the missing cc_toolchain features that are needed to try out clang modules (not c++ 20 modules)? It looks like there are a few -- use_header_modules, header_module_compile, header_module_codegen -- that I wasn't able to find in the toolchain config but exist in the java code. And with those features added, is it as simple as CC=clang and adding build --features=use_header_modules to .bazelrc to try it out?

I'm not sure whether it will just work, but here are the missing features that are in the internal toolchain:
https://gist.github.com/oquenchil/e6e39237b4c7b95b7b96396764b9a97c

**Comment by det** [comment]
Thanks for the missing features @oquenchil!

I will note the following line from the `header_module_compile` feature:
> "-Xcrosstool-module-compilation",

The string `crosstool-module-compilation` is not present in the Bazel or LLVM codebases. Is this likely to matter?

**Comment by det** [comment]
Another issue is that bazel is generating module maps with quoted module names which clang 14.0.1 does not seem to like when compiling a module map to a pcm (via `--features=header_modules`):
```
bazel-out/k8-fastbuild/bin/external/com_google_absl/absl/base/config.cppmap:1:8: error: expected a module name after 'module'
module "@com_google_absl//absl/base:config" {
```

**Comment by det** [comment]
Shouldn't all those "-Xclang=-<flag>" be `-Xclang`, "-<flag>"? My clang doesn't (14.0.1) seem to recognize the former.

**Comment by det** [comment]
So I took the new module's configuration you posted for a spin, but I am still running into the same problem as I was previously. It's not clear what to do about module maps for libc/libc++. You can merge them into the [module_map](https://bazel.build/reference/be/c-cpp#cc_toolchain.module_map) artifact for the toolchain, but it will never be compiled by bazel. You must compile module maps which contain any non textual headers so I am not sure how to make this work. Is it possible there is some internal Google secret sauce that is missing before anyone else can use this? Maybe related to `-Xcrosstool-module-compilation`?

**Comment by hlopko** [comment]
Yeah, there are a couple of things happening in the snippet @oquenchil shared.

1. This is an internal toolchain that uses a a couple of clang wrappers. Those handle things like `-Xclang=-<flag>` flags or implement behavior needed for Bazel, for example ensuring that action doing header parsing creates an output.
2. This snippet relies on having `module_maps` feature (I think [this one](https://cs.opensource.google/bazel/bazel/+/master:tools/cpp/unix_cc_toolchain_config.bzl;l=63) works but I haven't tried). You also very likely need working `layering_check` and `parse_headers` features. 
3. And IMHO the biggest problem is the high-level design of internal support for Clang modules, at least for now. It is not assumed that all targets in the repository will be compiled into modules. Instead, only a couple of hand-picked, high value targets are compiled into modules, and everything else is built in a traditional way. The flag `-fmodules-embed-all-files` tells Clang to bundle the whole transitive closure of dependencies of a target into a single module. This will grow quadratically if enabled for all targets, but works quite nice for high value targets that don't change often. I can't say if this model works well for non-Google users though, and it is certainly not something that I'd have expected from Bazel.
4. As @det points out, you need to manually create/generate a module map for your toolchain libraries. I might be wrong, but I think internally we don't actuallly generate a Clang module only for the C++ standard library, we generate them for some libraries that users commonly explicitly depend on, such as Abseil. 

I guess the point I'm trying to make is that there is still a lot of work to be done to make Clang modules work in Bazel in general, and the snippet helps a bit, but do not assume that what you need is within a hand's reach. For context, [there's](https://github.com/carbon-language/carbon-lang/blob/trunk/bazel/cc_toolchains/clang_cc_toolchain_config.bzl) a toolchain that Carbon folks use to build Carbon, and it doesn't support Clang modules, even though it is created by the group of engineers who helped to implement the support for Clang modules in Bazel internally.

**Comment by aaronmondal** [comment]
We just added highly experimental support for standard C++ modules in [rules_ll](https://github.com/eomii/rules_ll) (not the Clang variant with module maps). Considering how experimental C++ modules are currently in Clang/LLVM, things work quite well so far :sweat_smile: 

Build files look like this:
```python
ll_library(
    name = "mymodule",
    srcs = ["m.cpp"]  # Module implementation unit
    transitive_interfaces = {"m.cppm": "m"},  # Mapping module interface unit to module name
)

ll_binary(
  name = "main",
  srcs = ["main.cpp"],
  deps = [":mymodule"],
)
```
Docs with examples and on how things work under the hood:
https://ll.eomii.org/guides/modules.html

**Comment by rnburn** [comment]
@aaronmondal - with your rules is there any way for module libraries to depend on regular cc_library's? I see this in the [docs](https://ll.eomii.org/defs.html#ll_library-deps):
> Every dependency needs to be an ll_library.

**Comment by aaronmondal** [comment]
@rnburn We have some logic to reuse already-built cc_libraries from the llvm-project overlay, and at a glance I *think* putting a `cc_library` in the `llvm_project_deps` *may* work. There may be a need to specify custom include paths, but apart from that I don't think that this attribute is actually specific to LLVM targets.

Note that `ll_*` targets will use a different toolchain than the `cc_toolchain`. I think we can go from `cc_*` to `ll_*`, but it would probably be very hard to go back. If `cc_*` targets work, we'll rename the `llvm_project_deps` attribute :innocent:

**Comment by HappyCerberus** [comment]
There seems to be some progress on the Clang side: https://releases.llvm.org/15.0.0/tools/clang/docs/StandardCPlusPlusModules.html

Any corresponding changes on Bazel's side?

**Comment by aaronmondal** [comment]
@HappyCerberus This is the one we implemented in `rules_ll`. The current state is likely not stable enough for native rules. For instance, lambdas are not merged correctly in certain cases and we cannot use Clang's BMI caching system when building modules with Bazel.

Currently, BMIs contain absolute paths to their sources at the time of precompilation. This means that sandboxed precompilations are hard to implement and BMIs are currently not reproducible. We have not yet managed to get rid of these sandbox paths, so we work around this by disabling sandboxing for precompilations. This is not ideal and would not be viable in the native rules.

We didn't try with `libstdc++`, but `libcxx` still requires manual patching to work with global module fragments. This is another big limitation. Even if the native rules supported C++ modules, they would be essentially unusable since they use the host's C++ standard library.

So for C++ modules to be viable in native rules, distributions will need to distribute `libcxx` 16 or 17, assuming the current issues are fixed by then. Otherwise there would be infinite bug reports because not even `<iostream>` works without patches.

**Comment by nadiasvertex** [comment]
C++23 is here... any progress on this? I can't find anything definitive from Google.

**Comment by aaronmondal** [comment]
`rules_ll` upstream supports the `std` module from libc++, and Clang 17 will ship the `std` module as an experimental preview. Most issues have been fixed, and I'd consider modules as pretty much fully usable.

A few issues remain which I could imagine to be fairly hard blockers for `rules_cc`:

- Modules built with Clang aren't reproducible. Some absolute paths leak into BMIs. This means that you can't use them with remote execution. This is a tricky situation and a critical issue for Bazel because this can subtly poison your cache and might go unnoticed for a long time.
- BMIs aren't movable. Modules have to be distributed as `.cppm` files and they have to be built locally. If you depend on a modularized external dependency, you can't import a prebuilt BMI of that dependency. You have to build the entire dependency from source yourself. Not an issue if you do that anyways, but many users would probably prefer including some headers and linking a prebuild archive over building the entire dependency from source.
- There is no interop with modules and heterogeneous code. If you want to use modules and let's say CUDA in the same project, you need a very specific code layout to ensure that the CUDA code and module code never see each other during compilation and only get combined during linking.

We'll probably need distros to actually ship Clang 17 with the `std` module `.cppm` files to get a better feeling on what's good and bad for build-system UX with modules. The implementation in `rules_ll` certainly has some subtleties regarding inter-module visibility that I've reworked several times now and might have to rework several times more.

Then again C++23 is basically Rust already. Seems like that's also where all the Bazel engineering power is being redirected to. I wouldn't find it too surprising if the recommendation for the future is to just not write new C++ and use `rules_rust` with `crubit` for legacy-compatibility instead :smile:

**Comment by nadiasvertex** [comment]
That’s a shame. I was hoping to use Bazel for a large project we are migrating. C++20 modules are more important to me than  Bazels many advantages, and rules_ll is a non-starter because I need to support macOS and Windows.

Maybe next year things will be better.

**Comment by BorisChenCZY** [comment]
Do we have any update on this?

**Comment by BorisChenCZY** [comment]
I'm also quite interested in contributing this featuer. Is there any guidance on this for me to do so?

**Comment by nadiasvertex** [comment]
I would help with such an effort too.

**Comment by oquenchil** [comment]
See https://github.com/bazelbuild/bazel/issues/4005#issuecomment-1216281033

**Comment by nadiasvertex** [comment]
> See [#4005 (comment)](https://github.com/bazelbuild/bazel/issues/4005#issuecomment-1216281033)

It sounds like that is basically saying nothing is done and some one wanting good support should start from scratch.

**Comment by jsharpe** [comment]
Note also that post was referring to clang modules and not c++20 modules; this issue got hijacked from clang module support to c++ 20 at some point in the history of this issue.

**Comment by PikachuHyA** [comment]
Dear all, the support for C++20 Modules is here. 
PR: https://github.com/bazelbuild/bazel/pull/19940
design doc: https://github.com/bazelbuild/proposals/pull/354
discussion: https://github.com/bazelbuild/bazel/discussions/19939
a demo built with C++20 Modules (about 10 kloc): https://github.com/PikachuHyA/async_simple
extra tests: https://github.com/PikachuHyA/bazel_cxx20_module_test
the new bazel is built by GitHub Action, see https://github.com/PikachuHyA/bazel/actions/runs/6638058320
you can download the prebuilt version and play with your code.
Feedback is very welcome.
please note, only Clang is supported now. and Clang 17 or higher is required.

**Comment by kelteseth** [comment]
This issue is track it on https://arewemodulesyet.org/tools/

**Comment by PikachuHyA** [comment]
> Dear all, the support for C++20 Modules is here. PR: [#19940](https://github.com/bazelbuild/bazel/pull/19940) design doc: [bazelbuild/proposals#354](https://github.com/bazelbuild/proposals/pull/354) discussion: [#19939](https://github.com/bazelbuild/bazel/discussions/19939) a demo built with C++20 Modules (about 10 kloc): https://github.com/PikachuHyA/async_simple extra tests: https://github.com/PikachuHyA/bazel_cxx20_module_test the new bazel is built by GitHub Action, see https://github.com/PikachuHyA/bazel/actions/runs/6638058320 you can download the prebuilt version and play with your code. Feedback is very welcome. please note, only Clang is supported now. and Clang 17 or higher is required.

FYI: after https://github.com/bazelbuild/bazel/pull/22553 was merged, I created a repo demonstrating C++20 Modules with Bazel: https://github.com/PikachuHyA/bazel_cxx20_modules_demo/. 

It currently contains a simple hello‑world example and a demo showing how to use a third‑party library (e.g., fmt) with C++20 Modules.  

Feel free to open issues or submit PRs to add more examples.

**Comment by PikachuHyA** [comment]
hi @comius , could you please reopen this issue?

The current implementation for C++20 Modules support has significant limitations. We previously had a [solution](https://github.com/bazelbuild/bazel/pull/27927), but the patch was reverted https://github.com/bazelbuild/bazel/pull/28190 due to performance concerns. We’re still actively working on improving C++20 Modules support and would like to keep this issue open to track progress.
